internal/semantic/version-alpine.go

```diff
@@ -0,0 +1,338 @@
+	hash string
+	// prefixed with "-r{number}"
+	buildComponent *big.Int
+}
+
+func (v AlpineVersion) compareComponents(w AlpineVersion) int {
+	numberOfComponents := maxInt(len(v.components), len(w.components))
+
+	for i := 0; i < numberOfComponents; i++ {
+		diff := v.components.Fetch(i).Cmp(w.components.Fetch(i))
+
+		if diff != 0 {
+			return diff
+		}
+	}
+
+	return 0
+}
+
+func (v AlpineVersion) compareLetters(w AlpineVersion) int {
+	if v.letter == "" && w.letter != "" {
+		return -1
+	}
+	if v.letter != "" && w.letter == "" {
+		return +1
+	}
+
+	return strings.Compare(v.letter, w.letter)
+}
+
+func (v AlpineVersion) fetchSuffix(n int) alpineSuffix {
+	if len(v.suffixes) <= n {
+		return alpineSuffix{number: big.NewInt(0), weight: 5}
+	}
+
+	return v.suffixes[n]
+}
+
+func (as alpineSuffix) Cmp(bs alpineSuffix) int {
+	if as.weight > bs.weight {
+		return +1
+	}
+	if as.weight < bs.weight {
+		return -1
+	}
+
+	return as.number.Cmp(bs.number)
+}
+
+func (v AlpineVersion) compareSuffixes(w AlpineVersion) int {
+	numberOfSuffixes := maxInt(len(v.suffixes), len(w.suffixes))
+
+	for i := 0; i < numberOfSuffixes; i++ {
+		diff := v.fetchSuffix(i).Cmp(w.fetchSuffix(i))
+
+		if diff != 0 {
+			return diff
+		}
+	}
+
+	return 0
+}
+
+func (v AlpineVersion) compareBuildComponents(w AlpineVersion) int {
+	if v.buildComponent != nil && w.buildComponent != nil {
+		if diff := v.buildComponent.Cmp(w.buildComponent); diff != 0 {
+			return diff
+		}
+	}
+
+	return 0
+}
+
+func (v AlpineVersion) compareRemainder(w AlpineVersion) int {
+	if v.remainder == "" && w.remainder != "" {
+		return +1
+	}
+
+	if v.remainder != "" && w.remainder == "" {
+		return -1
+	}
+
+	return 0
+}
+
+func (v AlpineVersion) Compare(w AlpineVersion) int {
+	// if both versions are invalid, then just use a string compare
+	if v.invalid && w.invalid {
+		return strings.Compare(v.original, w.original)
+	}
+
+	// note: commit hashes are ignored as we can't properly compare them
+	if diff := v.compareComponents(w); diff != 0 {
+		return diff
+	}
+	if diff := v.compareLetters(w); diff != 0 {
+		return diff
+	}
+	if diff := v.compareSuffixes(w); diff != 0 {
+		return diff
+	}
+	if diff := v.compareBuildComponents(w); diff != 0 {
+		return diff
+	}
+	if diff := v.compareRemainder(w); diff != 0 {
+		return diff
+	}
+
+	return 0
+}
+
+func (v AlpineVersion) CompareStr(str string) int {
+	return v.Compare(parseAlpineVersion(str))
+}
+
+// parseAlpineNumberComponents parses the given string into AlpineVersion.components
+// and then returns the remainder of the string for continued parsing.
+//
+// Each number component is a sequence of digits (0-9), separated with a ".",
+// and with no limit on the value or amount of number components.
+//
+// This parser must be applied *before* any other parser.
+func parseAlpineNumberComponents(v *AlpineVersion, str string) string {
+	sub := cachedregexp.MustCompile(`^((\d+)\.?)*`).FindString(str)
+	str = strings.TrimPrefix(str, sub)
```

Comment:
nit: move this down directly to the return statement, it was a bit confusing to me up here as I thought the result was used in the loop.